[Zhaoqi] iP

Eclipse-Dominator opened the pull request:
# Duke 

A magical checklist application

------



## what does duke have to offer

- Magical yet mysterious wonders

- ~~Hard~~  **Easy**  to use

- Tonnes of features

  ------

  

## what are you waiting for?

1. Download the [future](https://www.youtube.com/watch?v=dQw4w9WgXcQ) now!
2. Please don't click on the link above, you *will* regret it.

>  "Yup, best of all, it's **Ad-free**"
>
> "Wonderful, what brilliancy!"

​				~Users

Features:

- [ ] CLI pattern highlighting (only works on cli and not gui sadly)
- [x] Add Todo Task
- [x] Add Deadline Task
- [x] Add Event Task
- [x] Automatic datetime parsing if input format is datetime

------

There is even some python code to generate Test files too!

```Python
sample_header = '''
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.assertEquals;
public class {} {{
}}
'''

source_folder = "./src/main/java/"
test_folder = "./src/test/java/"


for root, dirs, files in os.walk(source_folder):
    curr_path_in_test = f"{test_folder}{root[len(source_folder):]}"
    print(f"generating dir at {curr_path_in_test}")
    for dir in dirs:
        curr_path = os.path.join(curr_path_in_test, dir)
        if not os.path.isdir(curr_path):
            os.mkdir(curr_path)
     # ....
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
updated gitignore

Eclipse-Dominator committed:
file update?

Eclipse-Dominator committed:
Removed carriage return to work with bash

Eclipse-Dominator committed:
Added base and IO class

Eclipse-Dominator committed:
Complete Level-1

Eclipse-Dominator committed:
- Added outro text on exit
- Completed Level 2

Eclipse-Dominator committed:
Added parser to split space and /

Eclipse-Dominator committed:
Added base task class

Eclipse-Dominator committed:
ParsedData to be used to transfer data

Eclipse-Dominator committed:
Fixed missing formatting

Eclipse-Dominator committed:
dynamic array and support for commands

Eclipse-Dominator committed:
update task

Eclipse-Dominator committed:
Added child class deadline, event and todo

Eclipse-Dominator committed:
renamed some variables

Eclipse-Dominator committed:
Updated main method to support the new tasks type

Eclipse-Dominator committed:
Fixed some formatting

Eclipse-Dominator committed:
Added basic temporary testing

Eclipse-Dominator committed:
Added custom DukeExceptions

Eclipse-Dominator committed:
Implemented the use of Custom exceptions

Eclipse-Dominator committed:
renamed invalid class name

Eclipse-Dominator committed:
Updated testing to handle the new errors

Eclipse-Dominator committed:
Updated exception message

Eclipse-Dominator committed:
added delete and updated exception handling

Eclipse-Dominator committed:
Updated testing to support for delete operation

Eclipse-Dominator committed:
Remove raw field from parsed data

Eclipse-Dominator committed:
added basic file management

Eclipse-Dominator committed:
added functionality to convert task to save string

Eclipse-Dominator committed:
updated .ignore to ignore user save data

Eclipse-Dominator committed:
Changed exception signature

Eclipse-Dominator committed:
Added method to handle saving and reading

Eclipse-Dominator committed:
Abstracted out task into a seperate class

Eclipse-Dominator committed:
Fixed rewrite bug for appending to file

Eclipse-Dominator committed:
Added regex matcher to capture info from save

Eclipse-Dominator committed:
Integrated saving of data to duke

Eclipse-Dominator committed:
Added auto creation of folders

Eclipse-Dominator committed:
Added generation of folders

Eclipse-Dominator committed:
Merge branch 'branch-level-7' of https://github.com/Eclipse-Dominator/ip into branch-Level-7

Eclipse-Dominator committed:
Removed erronous code

Eclipse-Dominator committed:
updated gitignore and storage

Eclipse-Dominator committed:
Fixed creation of folders

Eclipse-Dominator committed:
Implemented datetime into event and deadline

Eclipse-Dominator committed:
Merge branch 'branch-Level-7'

Eclipse-Dominator committed:
Merge main into branch-Level-8

Eclipse-Dominator committed:
Make tasks sortable

Eclipse-Dominator committed:
Remove  unused imports

Eclipse-Dominator committed:
Add filter task before given date

Eclipse-Dominator committed:
Merge branch 'branch-Level-8'

Eclipse-Dominator committed:
Added packages

Eclipse-Dominator committed:
Update file name to fit naming std

Eclipse-Dominator committed:
Abstract Command package

Eclipse-Dominator committed:
Integrate Commands into main

Eclipse-Dominator committed:
Merge remote-tracking branch 'Upstream-HEAD/add-gradle-support'

Eclipse-Dominator committed:
Remove dependencies

Eclipse-Dominator committed:
Restructure files

Eclipse-Dominator committed:
updated gitignore

Eclipse-Dominator committed:
Add a method to generate Class Tests

Eclipse-Dominator committed:
Switch to newer gradle syntax and version

Eclipse-Dominator committed:
Fix NoClassDefFoundError due to dependency

Eclipse-Dominator committed:
Generate outline of all Test files

Eclipse-Dominator committed:
Add test case for Todo

Eclipse-Dominator committed:
Add Test for ParsedData

Eclipse-Dominator committed:
Add partial testing for Parser

Eclipse-Dominator committed:
Update shadowJar ver to work with gradle 7.x.x+

Eclipse-Dominator committed:
fix untracked files

Eclipse-Dominator committed:
Add javadoc for all commands

Eclipse-Dominator committed:
Add javadoc to exceptions

Eclipse-Dominator committed:
Add Javadoc for Duke

Eclipse-Dominator committed:
Add javadoc for tasks

Eclipse-Dominator committed:
Update Javadoc for ParsedData

Eclipse-Dominator committed:
Add javadoc for the util classes

Eclipse-Dominator committed:
Add text highlighting and Find command

Eclipse-Dominator committed:
Merge branch 'branch-A-JavaDoc'

Eclipse-Dominator committed:
Merged branch 'branch-Level-9'

Eclipse-Dominator committed:
Fixed some Javadoc after mege conflict

Eclipse-Dominator committed:
Merge branch 'master' into branch-A-CodingStandard

Eclipse-Dominator committed:
Reformmatted indents

Eclipse-Dominator committed:
Merge branch 'branch-A-CodingStandard'

Eclipse-Dominator committed:
Add checkstyle to gradle

dexter-sim commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,116 @@
+    private static final String INTRO = "Hello! I'm Duke\n"
+            + "What can I do for you?";
+
+    private DukeIo userIO;
```
The textbook states "Abbreviations and acronyms should not be uppercase when used as a part of a name.".
Do you think it is better to name it `userIo` or `userInputOutput`?

Eclipse-Dominator replied:
Updated.

dexter-sim commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    /**
+     * Initiate the running loop. Will exit when <code>handleInput</code> returns false.
```
The textbook states "In method header comments, the first sentence should start in the form `Returns ...`, `Sends ...`, `Adds ...` (not `Return` or `Returnning` etc.)".
Should this line start with `Initiates` instead?

Eclipse-Dominator replied:
updated.

dexter-sim commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,72 @@
+            throw new EmptyDescriptionException("deadline");
+
+        if (data.additionalInfo.length() == 0)
+            throw new EmptyTimeException("deadline", SPLIT);
```
The textbook states "Single statement conditionals should still be wrapped by curly brackets." and "Use K&R style brackets (aka [Egyptian style](https://blog.codinghorror.com/new-programming-jargon/)).".
Would you consider adding curly braces for these 2 single statement conditionals?

Eclipse-Dominator replied:
definitely, I had made the changes for this one but forgot to push it.

dexter-sim commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,97 @@
+ */
+public abstract class Task implements Comparable<Task> {
+    protected final String description;
+    protected boolean completed;
```
The textbook states "Boolean variables/methods should be named to sound like booleans.". 
Do you think `hasComplete` or `isComplete` would be more suitable name for a boolean variable?

Eclipse-Dominator replied:
updated to fit coding standard.

dexter-sim commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,97 @@
+     * 
+     * @return converted Task
+     */
+    abstract public ParsedData convertToParseData();
```
The textbook states "Method modifiers should be given in the following order: 
`<access> static abstract synchronized <unusual> final native`".
Would you agree that it should be ```public abstract ParsedData convertToParseData();``` instead?

Eclipse-Dominator replied:
definitely, I had made the changes for this one but forgot to push it.

dexter-sim commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,50 @@
+
+        io.printLine();
+        io.printTask(FIND_TXT, 0);
+        String sBefore, sAfter;
+        for (int i = 0; i < tasks.size(); i++) {
+            sBefore = tasks.get(i).toString();
+            sAfter = sBefore.replaceAll(searchString, RED_TXT);
```
The textbook states "Variables should be initialized where they are declared and they should be declared in the smallest scope possible."
Do you think that the declaration of the 2 strings be in the for loop instead of before the for loop?

Eclipse-Dominator replied:
Yup, moved the declaration to within the for loop.

dexter-sim reviewed (commented):
Looks good.
Some minor coding standard violations.

ElijahS67 commented on src/main/java/duke/command/ByCommand.java:
```diff
@@ -0,0 +1,49 @@
+     * @param d ParsedData containing a possible datetime pattern
+     */
+    public ByCommand(ParsedData d) {
+        super(d);
```
Perhaps a more descriptive name rather than "d" is better?

Eclipse-Dominator replied:
changed d -> parsedData.
Although for under the scope of the constructor which is really short, I think this is ok 

>  Variables with a large scope should have long names, variables with a small scope can have short names.

ElijahS67 commented on src/main/java/duke/command/TodoCommand.java:
```diff
@@ -0,0 +1,46 @@
+
+    private static final String ADD_TASK = "Got it. I've added this task:%n"
+            + "  %s%n"
+            + "Now you have %d tasks in the list.";
```
Perhaps the indentation of line 20 should be 8 spaces more than the parent line?

Eclipse-Dominator replied:
Line 20's parent line should still be line 18, so I don't think there should another 8 spaces on line 20. 
I think the parent line should change when there is a change in scope. 
ie 
```
demoSomeRandomMethodThatIsLong()
        .method1()
        .method2(demo2        // <-- no change in parent line
                .method3());        // <-- change in parent line
```

ElijahS67 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,97 @@
+     * @return a task is bigger if its time higher and biggest if non-existent
+     */
+    @Override
+    public int compareTo(Task o) {
```
Perhaps a more descriptive name than "o" should be used?

Eclipse-Dominator replied:
changed `o` -> `task`.
Although for under the scope which is short, I think this keeping it as o is fine

>  Variables with a large scope should have long names, variables with a small scope can have short names.

ElijahS67 reviewed (commented):
Looks good to merge!

Eclipse-Dominator committed:
Rename variables to match coding standard

Eclipse-Dominator committed:
add basic structure for javafx

Eclipse-Dominator committed:
Abstract out DukeIo as an interface

Eclipse-Dominator committed:
Add java fxml to gradle

Eclipse-Dominator committed:
Abstract more IO methods and string parsers

Eclipse-Dominator committed:
Add support for javafx

Eclipse-Dominator committed:
Reduce line length

Eclipse-Dominator committed:
Merge branch 'branch-Level-10'

Eclipse-Dominator committed:
enable assertions in gradle settings

Eclipse-Dominator committed:
Improve clarity of IO settings with enum

Prevously the format of the text: whether it is indented,
with a wrapper or nothing are handled using binary bit flags.
However, there was no enum to replace the bits.
This makes others difficult to understand what each bit mean.

The binary bit flag values are replaced using Enums
for much clearer readability and comprehension.

Eclipse-Dominator committed:
Abstract out make delay & exit app to follow SLAP

Eclipse-Dominator committed:
Abstract exit program to follow SLAP

Eclipse-Dominator committed:
Abstract out read and display user input

Eclipse-Dominator committed:
Add enum to replace tasks

Eclipse-Dominator committed:
Seperate commands enum from parser

Eclipse-Dominator committed:
Create new branch for additions of assertions

Eclipse-Dominator committed:
Merge pull request #3 from Eclipse-Dominator/branch-A-Assertions

Create new branch for additions of assertions

Eclipse-Dominator committed:
Merge pull request #2 from Eclipse-Dominator/branch-A-CodingStandard

A-CodeQuality PR request

Eclipse-Dominator committed:
remove invalid assertion

Eclipse-Dominator committed:
Add missing javadoc for StringParser

Eclipse-Dominator committed:
Update files to fix checkstyle

Eclipse-Dominator committed:
Fix checkstyle issues

Eclipse-Dominator committed:
Set up CI

Eclipse-Dominator committed:
Adds step to make gradle executable

Eclipse-Dominator committed:
Adds ability to add custom alias for commands

Eclipse-Dominator committed:
Merge pull request #4 from Eclipse-Dominator/branch-C-FriendlierSyntax

Adds ability to add custom alias for commands

Eclipse-Dominator committed:
Abstract out getting tasks before xxx time

Eclipse-Dominator committed:
Abstract out magic string

Eclipse-Dominator committed:
Switch to adopt use of enum bit flags

Eclipse-Dominator committed:
Abstract out comparison with a enum bitflag

Eclipse-Dominator committed:
Fixes trailing whitespace issue in javadocs

Eclipse-Dominator committed:
Change in wrapping

Eclipse-Dominator committed:
Merge pull request #5 from Eclipse-Dominator/branch-A-CodeQuality

LGTM

Eclipse-Dominator committed:
Makes GUI resizable and changed bg color

Eclipse-Dominator committed:
Adds dialogue box for dialogues with shadow

Eclipse-Dominator committed:
Adds quality of life changes to GUI
- Change icon to circule
- Reduced icon size
- Used a flat color theme
- Added dialog boxes
- Red dialog box for warnings and error
- Added drop shadows and rounded corners
- Resizable GUI
- Includes a monospace font to display ascii art

Eclipse-Dominator committed:
Adds duo mode and file reading.
- Launch with GUI or without.
- Ability to choose save file when launching
- Custom generated CLI help screen

Eclipse-Dominator committed:
Update to match checkstyle

Eclipse-Dominator committed:
Update to use a more vibrant color theme

Eclipse-Dominator committed:
Merge pull request #6 from Eclipse-Dominator/branch-A-BetterGui

Branch a better gui

Eclipse-Dominator committed:
Adds method to print unordered list

Eclipse-Dominator committed:
Fixes aliasing bug for Deadline and Event

Eclipse-Dominator committed:
Removes dead code

Eclipse-Dominator committed:
Renames some javadoc to fit coding standard

Eclipse-Dominator committed:
Changes some text and adds a way to swap icons

Eclipse-Dominator committed:
Merge pull request #7 from Eclipse-Dominator/branch-A-Personality

Adds personality to Duke

Eclipse-Dominator committed:
Adds exception when cli access a gui feature

Eclipse-Dominator committed:
Adds exception when image failed to download

Eclipse-Dominator committed:
Adds exception when multiple async tasks are ran

Eclipse-Dominator committed:
Adds more error handling to update face command

Eclipse-Dominator committed:
Merge pull request #8 from Eclipse-Dominator/branch-A-MoreErrorHandling

Branch a more error handling

Eclipse-Dominator committed:
Adds unit test for util and task

Eclipse-Dominator committed:
Removes storageTest

Eclipse-Dominator committed:
Adds unit test for command selector.

Eclipse-Dominator committed:
Adds CLII auto testing

Eclipse-Dominator committed:
Adds shortcut to reset all aliases

Eclipse-Dominator committed:
Updates github actions

Eclipse-Dominator committed:
Updates gitignore

Eclipse-Dominator committed:
Updates CI

Eclipse-Dominator committed:
Fix some junit test

Eclipse-Dominator committed:
Merge pull request #9 from Eclipse-Dominator/branch-A-MoreTesting

Branch a more testing

Eclipse-Dominator committed:
Adds UG (imcomplete)

Eclipse-Dominator committed:
Merge branch 'master' of https://github.com/Eclipse-Dominator/ip

Eclipse-Dominator committed:
Updates UG readme.md

Eclipse-Dominator committed:
Updates UG readme.md

Eclipse-Dominator committed:
Merge branch 'master' of https://github.com/Eclipse-Dominator/ip

Eclipse-Dominator committed:
Finish UG.

Eclipse-Dominator committed:
Fix alias accepting empty  string

Eclipse-Dominator committed:
Updates UG to use ToC

Eclipse-Dominator committed:
Updates table to display properly

Eclipse-Dominator committed:
Updates table to display properly

Eclipse-Dominator committed:
Fix github page tabel

Eclipse-Dominator committed:
Update table style

Eclipse-Dominator committed:
Adds a custom theme

Eclipse-Dominator committed:
Changes table format

Eclipse-Dominator committed:
Fix table

Eclipse-Dominator committed:
Updates to YML

Eclipse-Dominator committed:
Update readme and table

Eclipse-Dominator committed:
Update yml config

Eclipse-Dominator committed:
Update Config and html

Eclipse-Dominator committed:
Fix default page to README.md

Eclipse-Dominator committed:
Remove unused header

Eclipse-Dominator committed:
Adds missing header in UG

Eclipse-Dominator committed:
Adds link to release file

Eclipse-Dominator committed:
Fixes some bugs
- Throws error when no parameter commands are called with command
- Fixes program not exiting when a fatal error occurs when running
- Allows save data to be created through nested folders
- Fixed notification of when parsing corrupted lines
